File: cp_rust/src/bin/cses-1068-weird-algorithm.rs
// Created by Ayush Biswas at 2025/07/08 11:12
// https://cses.fi/problemset/task/1068/
use cp_lib::*;

// @code begin
use cpio::*;
use std::{
    iter::once,
    ops::{AddAssign, DivAssign, MulAssign},
};

sol! {
    fn solution(
        n: usize,
    ) -> Words<usize> {
        let res = (1..).scan(n, |state, _| {
            if *state == 1 {
                return None;
            }

            if *state%2 == 0 {
                state.div_assign(2);
            } else {
                state.mul_assign(3);
                state.add_assign(1);
            }
            Some(*state)
        });

        once(n).chain(res).collect()
    }
}

// @code end

File: cp_haskell/bin/cses-1068-weird-algorithm.hs
-- Created by Ayush Biswas at 2026/03/03 15:21
-- https://cses.fi/problemset/task/1068

-- @code begin

import Text.Printf (printf)

weird xs@(1 : _) = xs
weird xs@(x : _) | even x = weird $ (x `div` 2) : xs
weird xs@(x : _) | odd x = weird $ (x * 3 + 1) : xs

main :: IO ()
main = do
  n :: Int <- read <$> getLine
  let res = reverse $ weird [n]
  mapM_ (printf "%d ") res

-- @code end